### PR TITLE
Bratseth/smarter rebalancing

### PR DESCRIPTION
**Pull request by bratseth**
Please look at each commit and read the commit messages.

### CODE REVIEW

**Review by freva — APPROVED** [review]
LGTM, but we should keep the lock for the entire duration of mark wantToRetire, prepare, activate(, unmark wantToRetire)

**Comment by bratseth** [comment]
It's a bit more ugly but ok, done. PTAL

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/Rebalancer.java [hunk 1]
```diff
@@ -90,20 +107,50 @@ private Move findBestMove(NodeList allNodes) {
...
-    private void markWantToRetire(Node node) {
+    private boolean markWantToRetire(Node node, boolean wantToRetire) {
         try (Mutex lock = nodeRepository().lock(node)) {
             Optional<Node> nodeToMove = nodeRepository().getNode(node.hostname());
-            if (nodeToMove.isEmpty()) return;
-            if (nodeToMove.get().state() != Node.State.active) return;
+            if (nodeToMove.isEmpty()) return false;
+            if (nodeToMove.get().state() != Node.State.active) return false;
+
+            if (node.status().wantToRetire() != wantToRetire)
```
Comment (freva) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/Rebalancer.java]: If they are equal, we should return `false` or at least make sure not to "undo" this in `finally` if we fail to achieve the wanted result.

Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/Rebalancer.java]: If it's already set to want to retire it's fine that we do it.
However, as you point out it is not fine to undo it if somebody else sets it and we failed (regardless of when it happens).

PTAL - I'm going to run out of "try" clauses soon.

Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/Rebalancer.java]: PTAL

**Review by freva — APPROVED** [review]
:100:

**Comment by bratseth** [comment]
Hello Merge Stop Enforcer

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/Rebalancer.java [hunk 2]
```diff
@@ -90,20 +107,52 @@ private Move findBestMove(NodeList allNodes) {
...
+        try (MaintenanceDeployment deployment = new MaintenanceDeployment(application, deployer, nodeRepository())) {
+            if ( ! deployment.isValid()) return false;
+
+            boolean couldMarkRetiredNow = markWantToRetire(move.node, true);
+            if ( ! couldMarkRetiredNow) return false;
+
+            try {
+                if ( ! deployment.prepare()) return false;
+                if (nodeRepository().getNodes(application, Node.State.reserved).stream().noneMatch(node -> node.hasParent(move.toHost.hostname())))
+                    return false; // Deployment is not moving the from node to the target we identified for some reason
```
Comment (freva) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/Rebalancer.java]: This leaves the reserved node with the application for 20-27 mins (until it is expired by `ReservationExpirer`), this is especially unfortunate in CD when we want to fail integration tests when some node are stuck in a non-active state.

We should prepare the application again with the node unmarked as wantToRetire or move it directly to dirty if that is safe.

FYI: @hmusum

Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/Rebalancer.java]: The latter should be fine here. After lunch ...

Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/Rebalancer.java]: https://github.com/vespa-engine/vespa/pull/11355

**Pull request merged**